File: Tree/num_trees.cpp
/*number of Trees : 트리의 개수를 세는 프로그램 작성*/
#include <bits/stdc++.h>
using namespace std;

const int MX = 502;
vector<int> adj[MX];
bool vis[MX];
bool isTree;
int n, m, u, v, trees;

// DFS 
void DFS(int cur, int prev){
    for (int nxt : adj[cur]){
        if (nxt == prev) continue;    // next와 부모일 경우 contniue
        if (vis[nxt]){
            isTree = false;           // 이미 방문한 곳을 방문하면 Connected Component가 트리가 아님을 의미
            continue;
        }
        vis[nxt] = true;
        DFS(nxt, cur);
    }
}

int main(void){
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    while(1){
        cin >> n >> m;
        if (!n && !m) break;    // n == 0 && m == 0 
        
        fill(vis, vis+n+1, 0);                      // visited list 초기화 
        for (int i=1; i<=n; i++) adj[i].clear();    // adj list(vector) 비우기
        trees = 0;
        
        while (m--){
            cin >> u >> v;
            adj[u].push_back(v);
            adj[v].push_back(u);
        }
        for (int i = 1; i <= n; i++){   // 모든 노드를 시작점으로 돌면서
            if (vis[i]) continue;
            vis[i] = true;
            isTree = true;
            DFS(i, -1);
            trees += isTree;            // +=-1
        }
        if (!trees) cout << "NO Trees." << '\n';
        else if (trees == 1) cout << "Only One Tree" << '\n';
        else cout << trees << " Trees." << '\n';
    }
}

File: Tree/num_trees.py
# Tree 개수 세기 : 모든 노드를 시작점으로 설정하여 돌면서 Count

num_node = 6
#edge = [(1,2),(2,3),(3,5),(4,5),(5,6)]
edge = [(1,2),(2,3),(1,3),(4,5),(5,6),(6,4)]

adj = [[] for i in range(num_node+1)]
for i in range(len(edge)):
    u, v = edge[i]
    adj[u].append(v) 
    adj[v].append(u)
    
#print("Tree adj list : {}".format(adj))

#p = [0]*(num_node+1)          # 부모노드(반드시 0 삽입) & visited list 불필요
vis = [False]*(num_node+1)     # 해당 노드를 방문여부 확인
# Tree DFS 탐색 
def DFS(curr, prev):
    global isTree           # global 함수 필요!!!
    for nxt in adj[curr]:
        if (nxt == prev): continue
        if (vis[nxt]):
            isTree = False  # Cycle이면 NO Tree!
            continue
        vis[nxt] = True
        DFS(nxt, curr)
        
# 모든 노드를 시작점으로 돌아야함
treeCnt = 0 
for i in range (1, num_node+1):
    if (vis[i]): continue
    vis[i] = True         # 시작점 방문표시 
    isTree = True   
    DFS(i, 0)            # DFS를 돌면서 Tree 여부 확인
    if (isTree): 
        treeCnt+=1

print(treeCnt)   
